pkg/controlsvc/controlsvc_test.go

```diff
@@ -0,0 +1,647 @@
+package controlsvc_test
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"io"
+	"net"
+	"os"
+	"testing"
+
+	"github.com/ansible/receptor/pkg/controlsvc"
+	"github.com/ansible/receptor/pkg/controlsvc/mock_controlsvc"
+	"github.com/ansible/receptor/pkg/logger"
+	"github.com/golang/mock/gomock"
+)
+
+const (
+	writeToConnError = "write to conn write message err"
+)
+
+func printExpectedError(t *testing.T, err error) {
+	t.Errorf("expected error %s", err)
+}
+
+func TestConnectionListener(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockNetceptor := mock_controlsvc.NewMockNetceptorForControlsvc(ctrl)
+	mockListener := mock_controlsvc.NewMockListener(ctrl)
+	logger := logger.NewReceptorLogger("")
+
+	connectionListenerTestCases := []struct {
+		name          string
+		expectedError bool
+		expectedCalls func(context.CancelFunc)
+	}{
+		{
+			name:          "return from context error",
+			expectedError: true,
+		},
+		{
+			name:          "error accepting connection",
+			expectedError: false,
+			expectedCalls: func(ctxCancel context.CancelFunc) {
+				mockListener.EXPECT().Accept().DoAndReturn(func() (net.Conn, error) {
+					ctxCancel()
+					return nil, errors.New("terminated")
+				})
+				mockNetceptor.EXPECT().GetLogger().Return(logger)
+			},
+		},
+	}
+
+	for _, testCase := range connectionListenerTestCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			ctx, ctxCancel := context.WithCancel(context.Background())
+			defer ctxCancel()
```

Comment:
Your calling `defer ctxCancel()` and then explicitly calling `ctxCancel()` in both cases, this dosnet seem needed. Consider removing the explicit calls and just keeping the `defer` one.